src/keys.rs

```diff
@@ -0,0 +1,329 @@
+// Copyright 2020 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fs::{read_dir, DirEntry, OpenOptions};
+use std::io::{self, Read, Write};
+use std::os::unix::ffi::OsStringExt;
+use std::os::unix::fs::PermissionsExt;
+use std::path::{Path, PathBuf};
+
+use error_chain::{bail, ChainedError};
+
+use crate::errors::{Result, ResultExt};
+
+const KEYS_SUBDIR: &str = "authorized_keys.d";
+
+pub(crate) fn read_keys(ssh_dir: &Path, out: &mut dyn Write, err: &mut dyn Write) -> Result<()> {
+    // check permissions on paths leading to KEYS_SUBDIR
+    ensure_safe_dir_mode(ssh_dir)?;
+    let authorized_keys_dir = ssh_dir.join(KEYS_SUBDIR);
+    ensure_safe_dir_mode(&authorized_keys_dir)?;
+
+    // read directory in lexical order
+    let mut entries = read_dir(&authorized_keys_dir)
+        .chain_err(|| format!("reading {}", authorized_keys_dir.display()))?
+        .collect::<Vec<_>>();
+    entries.sort_unstable_by(|a, b| {
+        a.as_ref()
+            .map(|ent| ent.path())
+            .unwrap_or_else(|_| PathBuf::new())
+            .cmp(
+                &b.as_ref()
+                    .map(|ent| ent.path())
+                    .unwrap_or_else(|_| PathBuf::new()),
+            )
+    });
+
+    // walk entries
+    for ent in entries {
+        // Report and ignore errors for individual files, so we don't lock
+        // the user out of their account.
+        match try_read_key_file(&authorized_keys_dir, ent, out) {
+            Ok(_) => (),
+            Err(e) => {
+                let _ = err.write_all(format!("{}\n", e.display_chain()).as_bytes());
+            }
+        };
+    }
+
+    Ok(())
+}
+
+fn ensure_safe_dir_mode(path: &Path) -> Result<()> {
+    let metadata = path
+        .metadata()
+        .chain_err(|| format!("couldn't stat {}", path.display()))?;
+    if !metadata.is_dir() {
+        bail!("not directory: {}", path.display());
+    }
+    let mode = metadata.permissions().mode() & 0o7777;
+    if mode != 0o700 {
+        bail!("bad permission on {}: {:04o} != 0700", path.display(), mode);
+    }
+    Ok(())
+}
+
+fn try_read_key_file(
+    dir_path: &Path,
+    ent: io::Result<DirEntry>,
+    out: &mut dyn Write,
+) -> Result<()> {
+    // unpack error
+    let ent = ent.chain_err(|| format!("reading {}", dir_path.display()))?;
+
+    // ignore dotfiles
+    if ent.file_name().into_vec()[0] == b'.' {
+        bail!("{} is a dotfile, ignoring", ent.path().display());
+    }
```

Comment:
Hmm, seems it'd be cleaner to do the filtering upfront in the caller before sorting.

Reply:
sshd ignores stderr from `AuthorizedKeysCommand`s, so the detailed error reporting is only useful for manual debugging.  For that purpose, I think it's useful to retain the "is a dotfile, ignoring" message, and it's convenient to put that in `try_read_key_file()` because the caller already has infrastructure for logging errors to stderr and continuing.